### ISSUE TITLE
is_fixed checks inside constraint.set_expr need to be changed

### ISSUE THREAD

**Issue by ghackebeil** [Pyomo/pyomo#30]
The checks for is_fixed when setting a constraint expression need to be changed to something more restrictive. However, is_constant is too restrictive. We need another expression classifier (is_numeric?) that checks for the existence of optimization variables, or even the possibility of them appearing in the future (e.g., because a mutable expression appears somewhere). Here is some pseudo code of a problematic case:

``` !python
y = variable()
x = variable()
expr = x == y
x.fix(1.0)
c = constraint(expr)
x.free()
```

Now c.lower and c.upper are equal to x (which are always supposed to be "numeric"), but there is nothing invalid about how this constraint was defined. It is simply a side-effect of our invalid code breaking apart the constraint expression in a way that doesn't make sense.

**Comment by jsiirola** [comment]
Isn't the correct solution to finally fix `Constraint.add()` so that it doesn't manipulate the expression that the user hands it?

That is, we defer all decisions of what the lower/upper bounds of a constraint are until the writer.  This becomes doubly nice after we rework the solver interface so that the solver owns the writer instance: since the writer will be the one manipulating the expressions, it will be able to inform the solver how to interpret (and map) things like duals back into the original model context.

**Comment by ghackebeil** [comment]
I agree that your solution would be better in the longer term. However, that also seems like a MUCH more extensive change (it would dramatically affect constraint.py, canonical_repn.py, and all of the solver plugins), leading me to doubt it will get done anytime soon.

I have use for the kind of expression classification that I mentioned outside of constraint.add, and this seems like a more realistic fix to a very long standing problem. I think it could be done in a day if we can decide on a name for this class of expressions.

**Comment by ghackebeil** [comment]
Fixed in #42
